### ISSUE TITLE
How to decrease val_mrcnn_mask_loss? (val_mrcnn_mask_loss is incresing instead of decreasing while mrcnn_mask_loss is decreasing)

### ISSUE THREAD

**Issue by sukruburakcetin** [matterport/Mask_RCNN#2718]
I'm training the model with the parameters:
Config file consists of:
```
{
    NAME = "Rooftops"
    IMAGES_PER_GPU = 1
    NUM_CLASSES = 1 + 1  # Background + Rooftops
    STEPS_PER_EPOCH = 100  
    IMAGE_SHAPE = [1024, 1024, 3]
    VALIDATION_STEPS = 50
}
```
Masks dimensions are 1024x1024 with overlapping 512 px. 

Training/valid dataset size = 936/234 -> 80% - 20%

This is my augmentation list:

```
augmentation = ia.augmenters.Sometimes(0.5, ia.augmenters.OneOf([ia.augmenters.Fliplr(0.5),
                                                                         ia.augmenters.Flipud(0.5),
                                                                         ia.augmenters.Affine(rotate=45)]))
```
This is how I train the layers, heads, and all:(I also load pretrained COCO weights at the beginning)

```
# Training - Stage 1
print("Training network heads")
model.train(dataset_train, dataset_val,
            learning_rate=config.LEARNING_RATE,
            epochs=40,
            layers='heads',
            augmentation=augmentation)

# Training - Stage 2
# Finetune layers from ResNet stage 4 and up
print("Fine tune Resnet stage 4 and up")
model.train(dataset_train, dataset_val,
            learning_rate=config.LEARNING_RATE,
            epochs=120,
            layers='4+',
            augmentation=augmentation)

# Training - Stage 3
# Fine tune all layers
print("Fine tune all layers")
model.train(dataset_train, dataset_val,
            learning_rate=config.LEARNING_RATE / 10,
            epochs=160,
            layers='all',
            augmentation=augmentation)
```
This is the loss graph:(still processing, I got a sample between the training session)
![alt text](https://github.com/sukruburakcetin/object-detection-on-aerial-images-using-maskrcnn/blob/main/loss.svg?raw=true)

However, the val loss graph is increasing instead of decreasing which seems overfitting.  (still processing, I got sample between training session)

![alt text](https://github.com/sukruburakcetin/object-detection-on-aerial-images-using-maskrcnn/blob/main/val_loss.svg?raw=true)

Does anyone know why validation loss increasing? Is there any way to decrease it?

Here is my output when I try to predict. It seems while bboxes are perfectly aligned, masks are barely filled. What I need is masks instead of bbox areas? Does anyone one how to solve this problem? I really appreciate it.

![alt text](https://github.com/sukruburakcetin/object-detection-on-aerial-images-using-maskrcnn/blob/main/indir%20(2).png?raw=true)

**Comment by nataliameira** [comment]
Hello @sukruburakcetin ,

You can try:
- decrease the learning rate;
- increase the number of epochs.

You can also try customizing the hyperparameters for your problem. Look at this: https://medium.com/analytics-vidhya/taming-the-hyper-parameters-of-mask-rcnn-3742cb3f0e1b

I hope these suggestions help you!

**Comment by sukruburakcetin** [comment]
Hello @nataliameira, first of all, thank you for the contribution you provided. Do you mind if I ask the total of 320 epochs isn't enough for training?

Actually, I tried decreasing LR with a different training scheme without any augmentation, it works but I still didn't get the actual increase in val_loss and especially val_mrcnn_mask_loss after epoch 10. What I want to do, discriminating rooftops' borders well and also the coordinates of rooftops masks are important for the special task I ride. When I get good output for masks, this time, I lose the rooftops' boundary detection sensitivity and get results like results in semantic segmentation.

May I ask, bboxes exist but there are no mask particles in some of the bboxes. How is this happening?

The link is very beneficial, thank you very much.

**Comment by nataliameira** [comment]
How did you annotate the images? Maybe it's a problem with your data.

I think 320 epochs should be enough. I also think that 936 images could provide an interesting result.

How many objects did you label? Are your labels polygons?

**Comment by sukruburakcetin** [comment]
First of all, I extracted binary masks from the vectorial information of the rooftops. After that, I used an algorithm that I customized personally and extracted the co-ordinates of vertices of the given binary masks that include rooftops. Then, I reallocated the co-ordinates as polygon points into the JSON file to imitate and get results like in the VIA annotator. Then, I feed the training with batches of raw images and a JSON file that includes polygon points of the given data. 

I tried to increase a total of 936 images by augmentation via using imgaug locally and physically before feeding the training and got worsened results. This made me think that my data annotation could have problems.

All of my masks have objects, there was only one mask that didn't include any object(rooftop) out of 936 .png images. If you want to examine my code that translates binary images into a json format file that includes polygons of the objects, you can examine it via this link: https://github.com/sukruburakcetin/binarymask2json

I can increase training data count but I want to get at least a desirable result from this small sample. I just want to know I'm doing the things right.

I really appreciate your enthusiasm, thanks in advance.

**Comment by nataliameira** [comment]
@sukruburakcetin 

I used the VIA annotator as well. I had no problems. I only had bad segmentation for small objects with few annotations. If I think of anything else I'll tell you. Good luck!!!

**Comment by sukruburakcetin** [comment]
Thank you very much @nataliameira. I'll try implementing my data by labelling manually with a small sample size to alternate and ease the possible pathways.
